<?xml version='1.0' encoding='utf-8'?>
<robot name="panda">
  
  <link name="panda_link0">
    <visual>
      <geometry>
        <mesh filename="../meshes/panda/link0.dae" />
      </geometry>
    </visual>
  </link>
  <link name="panda_link1">
    <visual>
      <geometry>
        <mesh filename="../meshes/panda/link1.dae" />
      </geometry>
    </visual>
  </link>
  
  
  <joint name="panda_joint1" type="revolute">
    <origin rpy="-2.6313949212650455e-06 -5.186632432689784e-08 -1.2780383157481656e-17" xyz="2.4693690209296155e-17 -4.567396885044367e-17 0.3330000000000001" />
    <parent link="panda_link0" />
    <child link="panda_link1" />
    <axis xyz="0 0 1" />
    <limit effort="87.0" lower="-2.8973" upper="2.8973" velocity="2.175" />
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973" />
  </joint>
  
  <link name="panda_link2">
    <visual>
      <geometry>
        <mesh filename="../meshes/panda/link2.dae" />
      </geometry>
    </visual>
  </link>
  
  <joint name="panda_joint2" type="revolute">
    <origin rpy="-1.5671631233681536 -0.0006063905110036522 -4.638746785761078e-16" xyz="-0.00016542285187714177 0.00046611959924426654 4.3408992507204626e-17" />
    <parent link="panda_link1" />
    <child link="panda_link2" />
    <axis xyz="0 0 1" />
    <limit effort="87.0" lower="-1.7628" upper="1.7628" velocity="2.175" />
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.7628" soft_upper_limit="1.7628" />
  </joint>
  
  <link name="panda_link3">
    <visual>
      <geometry>
        <mesh filename="../meshes/panda/link3.dae" />
      </geometry>
    </visual>
  </link>
  
  <joint name="panda_joint3" type="revolute">
    <origin rpy="1.5675434452162544 0.007116873693165347 -0.0005936954139923941" xyz="-4.018702287107603e-05 -0.31650707253508886 0.00046611688726827104" />
    <parent link="panda_link2" />
    <child link="panda_link3" />
    <axis xyz="0 0 1" />
    <limit effort="87.0" lower="-2.8973" upper="2.8973" velocity="2.175" />
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973" />
  </joint>
  
  <link name="panda_link4">
    <visual>
      <geometry>
        <mesh filename="../meshes/panda/link4.dae" />
      </geometry>
    </visual>
  </link>
  
  <joint name="panda_joint4" type="revolute">
    <origin rpy="1.5720164890105501 0.001378946498497616 -0.007051884242910425" xyz="0.08298433483169573 -0.0008116166118042272 0.0005086088803250237" />
    <parent link="panda_link3" />
    <child link="panda_link4" />
    <axis xyz="0 0 1" />
    <limit effort="87.0" lower="-3.0718" upper="-0.0698" velocity="2.175" />
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-3.0718" soft_upper_limit="-0.0698" />
  </joint>
  
  <link name="panda_link5">
    <visual>
      <geometry>
        <mesh filename="../meshes/panda/link5.dae" />
      </geometry>
    </visual>
  </link>
  
  <joint name="panda_joint5" type="revolute">
    <origin rpy="-1.569984004153357 0.003846076429680107 -0.0013253971011171648" xyz="-0.082508298986969 0.38458342771047327 0.000807558975830882" />
    <parent link="panda_link4" />
    <child link="panda_link5" />
    <axis xyz="0 0 1" />
    <limit effort="12.0" lower="-2.8973" upper="2.8973" velocity="2.61" />
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973" />
  </joint>
  
  <link name="panda_link6">
    <visual>
      <geometry>
        <mesh filename="../meshes/panda/link6.dae" />
      </geometry>
    </visual>
  </link>
  
  <joint name="panda_joint6" type="revolute">
    <origin rpy="1.5668251155588273 0.002949000686762804 0.0038450010406930564" xyz="-0.00015157481312079457 0.00035924901235692046 0.0005840719712579211" />
    <parent link="panda_link5" />
    <child link="panda_link6" />
    <axis xyz="0 0 1" />
    <limit effort="12.0" lower="-0.0175" upper="3.7525" velocity="2.61" />
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-0.0175" soft_upper_limit="3.7525" />
  </joint>
  
  <link name="panda_link7">
    <visual>
      <geometry>
        <mesh filename="../meshes/panda/link7.dae" />
      </geometry>
    </visual>
  </link>
  
  <joint name="panda_joint7" type="revolute">
    <origin rpy="1.571805285853835 3.401850246877615e-06 -0.002789392698002778" xyz="0.08620360708606715 -0.0016404761909481107 -0.00035750862524293586" />
    <parent link="panda_link6" />
    <child link="panda_link7" />
    <axis xyz="0 0 1" />
    <limit effort="12.0" lower="-2.8973" upper="2.8973" velocity="2.61" />
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973" />
  </joint>
  <link name="panda_link8" />
  <joint name="panda_joint8" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.107" />
    <parent link="panda_link7" />
    <child link="panda_link8" />
  </joint>
  <joint name="panda_hand_joint" type="fixed">
    <parent link="panda_link8" />
    <child link="panda_hand" />
    <origin rpy="0 0 -0.7853981633974483" xyz="0 0 0" />
  </joint>
  
  <link name="panda_hand">
    <visual>
      <geometry>
        <mesh filename="../meshes/panda/hand.dae" />
      </geometry>
    </visual>
  </link>
  <link name="panda_hand_tcp" />
  <joint name="panda_hand_tcp_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.1034" />
    <parent link="panda_hand" />
    <child link="panda_hand_tcp" />
  </joint>
  <link name="ball_link">
    <visual>
      <material name="red">
        <color rgba="1 0 0 1" />
      </material>

      <geometry>
        <sphere radius="0.005" />
      </geometry>
    </visual>
  </link>

  <joint name="ball_joint" type="fixed">
    <origin rpy="0.0 0.0 0.0" xyz="-0.000964425639282203 -0.007034307145014577 0.031774138437114886" />
    <parent link="panda_hand_tcp" />
    <child link="ball_link" />
  </joint>
  <link name="panda_leftfinger">
    <visual>
      <geometry>
        <mesh filename="../meshes/panda/finger.dae" />
      </geometry>
    </visual>
  </link>
  <link name="panda_rightfinger">
    <visual>
      <origin rpy="0 0 3.141592653589793" xyz="0 0 0" />
      <geometry>
        <mesh filename="../meshes/panda/finger.dae" />
      </geometry>
    </visual>
  </link>

    <joint name="panda_finger_joint1" type="prismatic">
    <parent link="panda_hand" />
    <child link="panda_leftfinger" />
    <origin rpy="0 0 0" xyz="0 0 0.0584" />
    <axis xyz="0 1 0" />
    <limit effort="100" lower="0.0" upper="0.04" velocity="0.2" />
    <dynamics damping="0.3" />
  </joint>
  <joint name="panda_finger_joint2" type="prismatic">
    <parent link="panda_hand" />
    <child link="panda_rightfinger" />
    <origin rpy="0 0 0" xyz="0 0 0.0584" />
    <axis xyz="0 -1 0" />
    <limit effort="100" lower="0.0" upper="0.04" velocity="0.2" />
    <mimic joint="panda_finger_joint1" />
    <dynamics damping="0.3" />
  </joint>
</robot>